<launch>

  <!-- These are the arguments you can pass this launch file, for example paused:=true -->
  <arg name="paused" default="false"/>
  <arg name="use_sim_time" default="true"/>
  <arg name="gui" default="true"/>
  <arg name="headless" default="false"/>
  <arg name="debug" default="false"/>
  <!-- <arg name="model" default="$(find robot_diff_drive)/urdf/robot_diff_drive.urdf.xacro"/> -->
  <arg name="model" default="$(find turtlebot3_description)/urdf/turtlebot3_burger.urdf.xacro"/>
  <arg name="rvizconfig" default="$(find robot_diff_drive)/rviz/urdf.rviz"/>
  <arg name="map_file" default="$(find robot_diff_drive)/maps/map_my_robot_house.yaml" />

  <arg name="first_goal" default="[-2.57, 0.607, 0.0]" />
  <arg name="second_goal" default="[-1.524, 1.468, 0.0]" />
  <arg name="third_goal" default="[-2.750, 1.786, 0.0]" />
  <arg name="battery_threshold" default="20" />
  <arg name="dock_station_pose" default="[-3.0, 2.0, 0.0]" />

  <arg name="node_start_delay" default="10.0" />

  <rosparam param="waypoints">[$(arg first_goal), $(arg second_goal), $(arg third_goal)]</rosparam>
  <rosparam param="times_to_repeat" subst_value="True">-1</rosparam>

  <rosparam param="first_goal" subst_value="True">$(arg first_goal)</rosparam>
  <rosparam param="second_goal" subst_value="True">$(arg second_goal)</rosparam>
  <rosparam param="third_goal" subst_value="True">$(arg third_goal)</rosparam>

  <rosparam param="battery_threshold" subst_value="True">$(arg battery_threshold)</rosparam>
  <rosparam param="dock_station_pose" subst_value="True">$(arg dock_station_pose)</rosparam>

  <rosparam param="tree_file" subst_value="True">$(find robot_diff_drive)/bt/tree.xml</rosparam>
  
  <include file="$(find turtlebot3_gazebo)/launch/turtlebot3_house.launch" />

  <include file="$(find robot_diff_drive)/launch/amcl.launch">
    <!-- <arg name="scan_topic" value="/laser/scan"/> -->
    <arg name="scan_topic" value="scan"/>
  </include>

  <include file="$(find robot_diff_drive)/launch/move_base.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <node pkg="map_server" type="map_server" name="map_server" args="$(arg map_file)" />

  <param name="robot_description" command="xacro $(arg model)" />

  <rosparam command="load" file="$(find robot_diff_drive)/config/joints.yaml" ns="joint_state_controller" />

  <!-- Before using above don't forget about shutdown sign -->
  <node name="robot_controller_spawner" pkg="controller_manager" type="spawner"
    args="joint_state_controller
          --shutdown-timeout 3" />

  <node pkg="robot_state_publisher" type="robot_state_publisher"  name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="30.0" />
  </node>

  <node pkg="robot_diff_drive" type="tree" name="tree" output="screen" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' "/>

  <!-- <node name="rqt_robot_steering" pkg="rqt_robot_steering" type="rqt_robot_steering">
    <param name="default_topic" value="/cmd_vel"/>
  </node> -->

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(arg rvizconfig)" />

</launch>